### PR TITLE
Log drop notification

### PR DESCRIPTION
**Pull request by nordic-krch**
Extended backend interface to allow notifying backend  that log messages has been dropped due to insufficient internal buffer size. Notification contains number of log messages dropped since last notification. It is optional for a backend to implement handler.

It touches #11763 for log core. Howeverm updates in backends are required to fix the issue.

### CODE REVIEW

**Review by pizi-nordic — APPROVED** [review]

**Comment by pfalcon** [comment]
Commit message typo:

> Howeverm

File: include/logging/log_backend.h [hunk 1]
```diff
@@ -94,7 +95,25 @@ static inline void log_backend_put(const struct log_backend *const backend,
...
+ *
+ * Function is optional.
+ *
+ * @param[in] backend  Pointer to the backend instance.
+ * @param[in] cnt      Number of dropped logs since last notification.
+ */
+static inline void log_backend_dropped(const struct log_backend *const backend,
+				       u32_t cnt)
+{
+	assert(backend);
```
Comment (jarz-nordic) [inline include/logging/log_backend.h]: Please use `__ASSERT_NO_MSG`

**Comment by pfalcon** [comment]
> It is optional for a backend to implement handler.

That's exactly why we have tickets titled "logging cannot be trusted". Why after all this discussion not make it mandatory? Careless programmer can still stub it with an empty function.

File: include/logging/log_backend.h [hunk 2]
```diff
@@ -94,7 +95,25 @@ static inline void log_backend_put(const struct log_backend *const backend,
 }
 
 /**
- * @brief Function for reconfiguring backend to panic mode.
+ * @brief Notify backend about dropped log messages.
+ *
+ * Function is optional.
```
Comment (pfalcon) [inline include/logging/log_backend.h]: To what exactly this "optional" note applies? This is docstring for log_backend_dropped(), it is not optional and should be called properly by the logging core.

File: include/logging/log_backend.h [hunk 3]
```diff
@@ -105,7 +124,7 @@ static inline void log_backend_panic(const struct log_backend *const backend)
 }
 
 /**
- * @brief Function for setting backend id.
+ * @brief Set backend id.
```
Comment (pfalcon) [inline include/logging/log_backend.h]: All these docstring changes should have gone as a separate commit.

**Review by jakub-uC — CHANGES_REQUESTED** [review]
Minor change needed

File: tests/subsys/logging/log_core/src/log_core_test.c [hunk 4]
```diff
@@ -25,6 +25,9 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME);
 
 typedef void (*custom_put_callback_t)(struct log_backend const *const backend,
 				      struct log_msg *msg, size_t counter);
+
+static bool in_panic;
```
Comment (pfalcon) [inline tests/subsys/logging/log_core/src/log_core_test.c]: Commit says it's "Add test for message drop notification", but suddenly, "in_panic" is added here.

File: tests/subsys/logging/log_core/src/log_core_test.c [hunk 5]
```diff
@@ -288,34 +301,6 @@ static void test_log_arguments(void)
 		      "Unexpected amount of messages received by the backend.");
 }
 
-/*
```
Comment (pfalcon) [inline tests/subsys/logging/log_core/src/log_core_test.c]: Such massive removals unrelated to the commit topic are really not ok.

**Review by pfalcon — COMMENTED** [review]
IMHO, some better structuring of changes would be helpful.

**Comment by pfalcon** [comment]
> Commit message typo:

>>    Howeverm

Ok, that was PR description, not commit message.

**Comment by pfalcon** [comment]
So, why not include a backend change in this PR?

File: tests/subsys/logging/log_core/src/log_core_test.c (hunk 4 above)
Comment (nordic-krch) [reply tests/subsys/logging/log_core/src/log_core_test.c]: in the commit i've reordered tests to put panic test at the bottom. Unfortunately, there is no cleanup  between test cases (e.g. reset) so logger state stays between tests and there is no going back when panic mode is started. flag in_panic is precaution for the future if someone adds test case after panic test case.

Comment (nordic-krch) [reply tests/subsys/logging/log_core/src/log_core_test.c]: ok, let me split that into 2 commits. first tests reordering and then adding new test cases.

**Comment by nordic-krch** [comment]
> That's exactly why we have tickets titled "logging cannot be trusted". Why after all this discussion not make it mandatory? Careless programmer can still stub it with an empty function.

Reasoning for optional:
- to allow us to introduce this feature backend by backend
- backward compatibility (someone may already implemented log own log backend)
- some backends may not be interested in handling that (e.g. flash backend).

Main reason is the first one. Once backends are aligned we can make it obligatory like other calls if we can leave with new requirement for backend interface (api breaker).

**Comment by nordic-krch** [comment]
@pfalcon 
> So, why not include a backend change in this PR?

I prefer to do it step by step. Bulk PRs tend to stuck.

File: tests/subsys/logging/log_core/src/log_core_test.c (hunk 4 above)
Comment (pfalcon) [reply tests/subsys/logging/log_core/src/log_core_test.c]: > ok, let me split that into 2 commits

Only if you find that not too cumbersome. And actually, perfectalize tests is not worth it, better to do it like that next time in such situation.

However, if you'd split commit to logging.h into docstring updates vs drop count support, I'd truly appreciate.

**Comment by pfalcon** [comment]
> I prefer to do it step by step. Bulk PRs tend to stuck.

That's true, and sounds good.

File: include/logging/log_backend.h (hunk 1 above)
Comment (nordic-krch) [reply include/logging/log_backend.h]: done, created dedicated commit for cleanup (doxygen + asserts)

File: tests/subsys/logging/log_core/src/log_core_test.c (hunk 4 above)
Comment (nordic-krch) [reply tests/subsys/logging/log_core/src/log_core_test.c]: log_backend.h cleanup extracted to separate commit.

**Review by pfalcon — APPROVED** [review]
Looks good, thanks for updates, let's move along with this, +1.

**Review by jakub-uC — APPROVED** [review]

**Pull request merged**